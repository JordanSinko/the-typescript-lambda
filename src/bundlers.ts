import { spawnSync, SpawnSyncOptions } from 'child_process';
import { BundlingOptions, ILocalBundling } from '@aws-cdk/core';
import { Runtime } from '@aws-cdk/aws-lambda';

export class LocalBundler implements ILocalBundling {
  //constructor(private readonly props: any) { }
  constructor() {}

  tryBundle(_outputDir: string, _options: BundlingOptions): boolean {
    const command = [
      'esbuild`',
      `index.ts`,
      `--outdir=/asset-output`,
      `--platform=node`,
      `--target=node10`,
      `--format=cjs`,
      `--bundle`,
      `--sourcemap`,
      `--minify`,
    ].join(' ');

    console.log(command);

    exec('bash', ['-c', command], {
      env: { ...process.env },
      stdio: ['ignore', process.stderr, 'inherit'],
    });

    return true;
  }
}

export class DockerBundler {
  public readonly bundlingOptions: BundlingOptions;

  //constructor(options: any) {
  constructor() {
    this.bundlingOptions = {
      image: Runtime.NODEJS_12_X.bundlingDockerImage,
    };
  }
}

function exec(binary: string, args: string[], options?: SpawnSyncOptions) {
  const buff = spawnSync(binary, args, options);

  if (buff.error) {
    throw buff.error;
  }

  if (buff.status !== 0) {
    if (buff.stdout || buff.stderr) {
      throw new Error(
        `[Status ${
          buff.status
        }] stdout: ${buff.stdout
          ?.toString()
          .trim()}\n\n\nstderr: ${buff.stderr?.toString().trim()}`
      );
    }

    throw new Error(`${binary} exited with status ${buff.status}`);
  }

  return buff;
}
