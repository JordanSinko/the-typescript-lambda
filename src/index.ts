import * as path from 'path';
import { Code, Function as AwsFunction, FunctionOptions, Runtime, RuntimeFamily } from '@aws-cdk/aws-lambda';
import { Construct, BundlingDockerImage } from '@aws-cdk/core';

const ESBUILD_VERSION = '0.7.14';

export interface EsbuildFunctionProps extends FunctionOptions {
  readonly projectRoot: string;
  readonly entryPoints: string[];
  readonly runtime?: Runtime;
  readonly handler?: string;
  readonly outdir?: string;
  readonly target?: string;
  readonly bundle?: boolean;
  readonly sourcemap?: boolean;
  readonly minify?: boolean;
  readonly awsSdkConnectionReuse?: boolean;
}

export class EsbuildFunction extends AwsFunction {

  constructor(scope: Construct, id: string, props: EsbuildFunctionProps) {

    if (props.runtime && props.runtime.family !== RuntimeFamily.NODEJS) {
      throw new Error('\'runtime\' must be NodeJS family');
    }

    const runtime = props.runtime ?? Runtime.NODEJS_12_X;
    const handler = props.handler ?? 'handler';
    const code = Code.fromAsset(props.projectRoot, {
      bundling: {
        image: BundlingDockerImage.fromAsset(path.join(__dirname, '../esbuild'), {
          buildArgs: {
            IMAGE: runtime.bundlingDockerImage.image,
            ESBUILD_VERSION,
          },
        }),
        command: ['bash', '-c', 'touch /asset-output/hello.txt'],
      },
    });

    super(scope, id, {
      ...props,
      runtime,
      handler: `index.${handler}`,
      code,
    });
  }

}
