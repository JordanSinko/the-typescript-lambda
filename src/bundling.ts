import * as path from 'path';
import { AssetCode, Code, Runtime } from '@aws-cdk/aws-lambda';
import { LocalBundler, DockerBundler } from './bundlers';

export interface EsbuildBaseOptions {
  /**
   * Bundle all dependencies into the output files
   *
   * @default true
   */

  /**
   * Bundle all dependencies into the output files
   *
   * @default false
   */
  readonly minify?: boolean;

  /**
   * Emit a source map
   *
   * @default false
   */
  readonly sourcemap?: boolean;

  /**
   * Excludes modules from the bundle
   *
   * @default ['aws-sdk']
   */
  readonly externals?: string[];
}

export interface EsbuildOptions extends EsbuildBaseOptions {
  /**
   * The directory which contains necessary files and dependencies
   */
  readonly rootdir: string;

  /**
   * The file to build
   */
  readonly entry: string;

  /**
   * The runtime that will run the file
   */
  readonly runtime: Runtime;

  /**
   * The directory to build to
   */
  readonly outdir: string;
}

export class Bundle {
  public static esbuild(options: EsbuildOptions): AssetCode {
    if (options.rootdir == null) {
      throw new Error("Please specify 'rootDir'");
    }

    const localBundler = new LocalBundler({
      rootdir: path.resolve(options.rootdir),
      entry: path.resolve(options.entry),
      runtime: options.runtime,
      minify: options.minify ?? false,
      sourcemap: options.sourcemap ?? false,
      externals: options.externals ?? ['aws-sdk'],
    });

    const dockerBundler = new DockerBundler();

    return Code.fromAsset(options.rootdir, {
      bundling: {
        local: localBundler,
        ...dockerBundler.bundlingOptions,
      },
    });
  }
}
