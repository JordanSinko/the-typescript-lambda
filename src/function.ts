import * as path from 'path';
import { Function as AwsFunction, FunctionOptions, Runtime, RuntimeFamily } from '@aws-cdk/aws-lambda';
import { Construct, AssetHashType } from '@aws-cdk/core';
import { Bundle, EsbuildBaseOptions } from './bundling';

//const ESBUILD_VERSION = '0.7.14';

export interface EsbuildFunctionProps extends FunctionOptions, EsbuildBaseOptions {

  /**
     * The directory which contains necessary files and dependencies
     */
  readonly rootdir: string;

  /**
   * The file to build relative to `rootdir`
   */
  readonly entry: string;

  /**
   * The runtime that will run the file
   * 
   * @default @default - `NODEJS_12_X`
   */
  readonly runtime?: Runtime;

  /**
   * The exported handler function name
   * 
   * @default 'handler'
   */
  readonly handler?: string;

  /**
   * The directory to build to
   * 
   * @default '/asset-output'
   */
  readonly outdir?: string;

}

export class EsbuildFunction extends AwsFunction {

  constructor(scope: Construct, id: string, props: EsbuildFunctionProps) {

    if (props.runtime && props.runtime.family !== RuntimeFamily.NODEJS) {
      throw new Error('\'runtime\' must be NodeJS family');
    }

    const runtime = props.runtime ?? Runtime.NODEJS_12_X;
    const handler = props.handler ?? 'handler';
    const outdir = props.outdir ?? AssetHashType.OUTPUT;

    const code = Bundle.esbuild({
      rootdir: props.rootdir,
      entry: path.resolve(props.entry),
      outdir,
      runtime,
    });

    super(scope, id, {
      ...props,
      runtime,
      handler: `index.${handler}`,
      code
    });
  }

}
