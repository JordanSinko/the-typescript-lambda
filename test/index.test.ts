import '@aws-cdk/assert/jest';

import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { SynthUtils } from '@aws-cdk/assert';
import { Runtime } from '@aws-cdk/aws-lambda';
import { App, Stack } from '@aws-cdk/core';

import { EsbuildFunction } from '../src';

test('sets global tags on stack and resources', () => {

  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'the-typescript-lambda-test-'));
  const tmpFile = 'handler.ts';
  const tmpEntrypoint = path.join(tmpDir, tmpFile);

  console.log(tmpDir);

  fs.writeFileSync(tmpEntrypoint, 'export const handler = async () => \'Hello, world\';');

  const app = new App();
  const stack = new Stack(app, 'Stack', {});

  new EsbuildFunction(stack, 'Function', {
    entryPoints: [tmpFile],
    projectRoot: tmpDir,
    runtime: Runtime.NODEJS_12_X,
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
