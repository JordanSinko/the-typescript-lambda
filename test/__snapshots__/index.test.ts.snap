// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sets global tags on stack and resources 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters3547e169d713bb1fb0a0285a19cf6a89dd115a135dc69ad71b3dbfae186eb54fArtifactHashDCC7EDE0": Object {
      "Description": "Artifact hash for asset \\"3547e169d713bb1fb0a0285a19cf6a89dd115a135dc69ad71b3dbfae186eb54f\\"",
      "Type": "String",
    },
    "AssetParameters3547e169d713bb1fb0a0285a19cf6a89dd115a135dc69ad71b3dbfae186eb54fS3Bucket07908928": Object {
      "Description": "S3 bucket for asset \\"3547e169d713bb1fb0a0285a19cf6a89dd115a135dc69ad71b3dbfae186eb54f\\"",
      "Type": "String",
    },
    "AssetParameters3547e169d713bb1fb0a0285a19cf6a89dd115a135dc69ad71b3dbfae186eb54fS3VersionKey2DCA3A8E": Object {
      "Description": "S3 key for asset version \\"3547e169d713bb1fb0a0285a19cf6a89dd115a135dc69ad71b3dbfae186eb54f\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "Function76856677": Object {
      "DependsOn": Array [
        "FunctionServiceRole675BB04A",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters3547e169d713bb1fb0a0285a19cf6a89dd115a135dc69ad71b3dbfae186eb54fS3Bucket07908928",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters3547e169d713bb1fb0a0285a19cf6a89dd115a135dc69ad71b3dbfae186eb54fS3VersionKey2DCA3A8E",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters3547e169d713bb1fb0a0285a19cf6a89dd115a135dc69ad71b3dbfae186eb54fS3VersionKey2DCA3A8E",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "FunctionServiceRole675BB04A",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "FunctionServiceRole675BB04A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
